move/upgradable/v3/upgradable/sources/upgradable.move

```diff
@@ -0,0 +1,97 @@
+// Change the fields of singleton all together, beaking backwards compatibility.
+module upgradable::upgradable {
+    use std::ascii::{Self, String};
+
+    use sui::dynamic_field;
+    use sui::event;
+
+    use version_control::version_control::{Self, VersionControl};
+
+    const VERSION: u64 = 2;
+    const FUNCTION_INDEX_SET: u8 = 0;
+    const FUNCTION_INDEX_GET: u8 = 1;
+    const FUNCTION_INDEX_SET_TAG: u8 = 2;
+    const FUNCTION_INDEX_GET_TAG: u8 = 3;
+
+    const EIncorrectVersion: u64 = 0;
+
+    public struct Singleton has key {
+        id: UID,
+    }
+
+    public struct DataV0 has store {
+        value: u64,
+        version: u64,
+        version_control: VersionControl,
+    }
+
+    public struct DataV1 has store {
+        value: u64,
+        version: u64,
+        version_control: VersionControl,
+        // Add another field here
+        tag: String,
+    }
+
+    public struct ValueSet has copy, drop {
+        value: u64,
+    }
+
+    entry fun upgrade(self: &mut Singleton) {
+        let DataV0 { value, version, version_control: _} = dynamic_field::remove(&mut self.id, 0);
+        assert!(version < VERSION, EIncorrectVersion);
+
+        let version_control = version_control::new(vector[
+            vector[],
+            vector[],
+            vector[
+                FUNCTION_INDEX_GET, FUNCTION_INDEX_SET, FUNCTION_INDEX_GET_TAG
+            ],
+        ]);
+
+        let data = DataV1 {
+            value,
+            version: VERSION,
+            version_control,
+            tag: ascii::string(b""),
+        };
+        dynamic_field::add(&mut self.id, 0, data);
+    }
+
+    fun borrow_data(self: &Singleton): &DataV1 {
```

Comment:
Can the index syntax be used to simplify this?
https://move-book.com/reference/index-syntax.html